### PR TITLE
fix: resolve two artifact removal issues in tessellate_extract_features

### PR DESCRIPTION
**Pull request by raylim**
## Summary

This PR fully fixes `GrandQCArtifactRemover` so it genuinely runs during `tessellate_extract_features`, and adds fine-grained control over which tissue classes are removed.

---

#### Issue 1 — `artifact_remover_fn` was never wired in

`SegConfig.remove_artifacts=True` was passed as a plain boolean via `**OmegaConf.to_container(cfg.seg_config)`, but `segment_tissue()` only accepts an `artifact_remover_fn` callable. The remover was never created. Result: a warning was logged and artifact removal was silently skipped on every run.

**Fix**: `_tessellate_and_filter()` now instantiates `GrandQCArtifactRemover` from the config flags and passes it as `artifact_remover_fn` to `segment_tissue()`.

#### Issue 2 — Seg-level thumbnail too coarse for GrandQC

`segment_tissue()` reads a thumbnail at the lowest pyramid level (typically 8–16 µm/px). `GrandQCArtifactRemover` has an internal guard that returns the mask unchanged when `img_mpp > max_input_mpp` (default 8.0 µm/px) — so artifact removal still no-oped even after Issue 1 was fixed.

**Fix**: `segment_tissue()` now checks `img_mpp >= max_input_mpp` and reads a finer pyramid level at or below the threshold. The prediction map is resized back to seg-level mask dimensions with `INTER_NEAREST`. The `>=` (not `>`) handles exact-equality edge cases (e.g. a 0.5 µm/px slide at 16× downsample is exactly 8.0 µm/px).

#### Issue 3 — GrandQC weights downloaded once per slide

`GrandQCArtifactRemover` was instantiated inside `_tessellate_and_filter()`, which is called per slide. Weights are re-downloaded from HuggingFace on each new instantiation — causing latency and occasional timeouts for later slides in a batch.

**Fix**: Added `_build_artifact_remover(cfg)` helper. Both `_main_single` and `_main_batch` create one shared instance and pass it through the call chain.

#### Issue 4 — Silent failure when artifact removal empties the tissue mask

When GrandQC removes all tissue (e.g. out-of-distribution slides, borderline MPP), `cv2.findContours` returns `hierarchy=None`, causing a silent `return None` in `segment_tissue()` — no log, no error.

**Fix**: After calling the artifact remover, the tissue survival fraction is computed. If more than 90% of tissue was removed, the pre-removal mask is used instead and a warning is logged. The previously silent `hierarchy=None` path now also logs a warning.

---

### New feature — Fine-grained class exclusion (`exclude_classes`)

Replaces the `remove_penmarks_only: bool` parameter with `exclude_classes: FrozenSet[int]`.

#### Public class constants

```python
from mussel.utils.artifact_removal import (
    CLASS_GLASS,          # 0 — glass / clear-slide background
    CLASS_NORMAL_TISSUE,  # 1 — normal stained tissue
    CLASS_BLOOD,          # 2 — blood / haemorrhage
    CLASS_NECROSIS,       # 3 — necrosis
    CLASS_PEN_MARKING,    # 4 — pen marking
    CLASS_FOLD,           # 5 — tissue fold
    CLASS_HOLE,           # 6 — hole / physical damage
    CLASS_BACKGROUND,     # 7 — non-tissue background
)
```

#### Preset exclusion sets

| Preset | Classes removed | Use case |
|---|---|---|
| `EXCLUDE_PENMARKS_ONLY` | `{4, 7}` | **Default** — safe for all tissue types |
| `EXCLUDE_FOLDS_AND_PENMARKS` | `{4, 5, 6, 7}` | Also removes folds and holes |
| `EXCLUDE_ALL_ARTIFACTS` | `{2,3,4,5,6,7}` | Aggressive — old default behavior |

#### Usage examples

```python
from mussel.utils.artifact_removal import (
    GrandQCArtifactRemover,
    EXCLUDE_PENMARKS_ONLY, EXCLUDE_FOLDS_AND_PENMARKS, EXCLUDE_ALL_ARTIFACTS,
    CLASS_BLOOD, CLASS_PEN_MARKING, CLASS_BACKGROUND,
)

# Conservative (default): pen marks + background only
remover = GrandQCArtifactRemover()

# Moderate: also remove folds and holes
remover = GrandQCArtifactRemover(exclude_classes=EXCLUDE_FOLDS_AND_PENMARKS)

# Aggressive: everything except normal tissue
remover = GrandQCArtifactRemover(exclude_classes=EXCLUDE_ALL_ARTIFACTS)

# Custom: blood and pen marks only
remover = GrandQCArtifactRemover(
    exclude_classes=frozenset({CLASS_BLOOD, CLASS_PEN_MARKING, CLASS_BACKGROUND})
)
```

#### Config-driven control via `SegConfig.artifact_exclude_classes`

```yaml
# nextflow.config / titan_validation.yaml
tiling:
  remove_penmarks: true          # enable remover
  artifact_exclude_classes:      # explicit class list overrides flag-based preset
    - 4   # pen marking
    - 5   # fold
    - 6   # hole
    - 7   # background
```

Resolution order in `_build_artifact_remover`:
1. `artifact_exclude_classes` list → converted to `frozenset` verbatim
2. `remove_artifacts=True` → `EXCLUDE_ALL_ARTIFACTS`
3. `remove_penmarks=True` only → `EXCLUDE_PENMARKS_ONLY`

#### Backward compatibility

`remove_penmarks_only: bool` still works but emits `DeprecationWarning`:
```
GrandQCArtifactRemover: 'remove_penmarks_only' is deprecated.
  remove_penmarks_only=True  → exclude_classes=EXCLUDE_PENMARKS_ONLY
  remove_penmarks_only=False → exclude_classes=EXCLUDE_ALL_ARTIFACTS
```

---

### Changed files

| File | Change |
|---|---|
| `mussel/utils/artifact_removal.py` | Public `CLASS_*` constants, preset `EXCLUDE_*` sets, `exclude_classes` param, `torch.isin()` inference, deprecation shim |
| `mussel/utils/segment.py` | MPP escalation (`>=`), fraction-based fallback guard (10% survival threshold), warning on silent `hierarchy=None` |
| `mussel/cli/tessellate.py` | Strip `artifact_exclude_classes` from `**seg_cfg` before passing to `segment_tissue()` |
| `mussel/cli/tessellate_extract_features.py` | `_build_artifact_remover()` helper, shared instance in `_main_single` / `_main_batch` |
| `mussel/cli/tessellate_extract_features_common.py` | `artifact_remover_fn` param on 3 functions; resolve `exclude_classes` from config flags |
| `tests/mussel/cli/test_tessellate.py` | 4 existing tests updated to new API; 4 new tests |
| `tests/mussel/utils/test_artifact_removal.py` | Updated to use `exclude_classes` directly |

---

### Tests

All 22 relevant tests pass:

| Test | Covers |
|---|---|
| `test_artifact_remover_fn_wired_when_remove_artifacts` | `EXCLUDE_ALL_ARTIFACTS` when `remove_artifacts=True` |
| `test_artifact_remover_fn_not_instantiated_when_flags_false` | No remover when both flags false |
| `test_artifact_remover_fn_penmarks_only` | `EXCLUDE_PENMARKS_ONLY` when only `remove_penmarks=True` |
| `test_artifact_remover_fn_external_instance_reused` | External instance not re-instantiated |
| `test_segment_tissue_artifact_mpp_escalation` | Finer level read when MPP below threshold |
| `test_segment_tissue_artifact_mpp_exact_equality_escalates` | `>=` edge case: exact MPP equality triggers escalation |
| `test_segment_tissue_artifact_removal_empty_mask_fallback` | All-zeros result → pre-removal mask used |
| `TestGrandQCArtifactRemoverCore.*` (10 tests) | Class exclusion logic, shape preservation, MPP rescaling |

### CODE REVIEW

File: tests/mussel/cli/test_tessellate.py [hunk 1]
```diff
@@ -86,3 +86,163 @@ def test_seg_config_seg_model_neural():
...
+            "remove_penmarks": False,
+        },
+        "vis_config": {},
+        "keep_intermediate_files": False,
+        "gpu_device_id": 0,
+        "use_gpu": False,
+        "batch_size": 8,
+        "num_workers": 0,
+        "gpu_device_ids": None,
+    })
```
[the remark this answers is not in the document]
Comment (raylim) [reply tests/mussel/cli/test_tessellate.py]: Fixed in 84bfb20 — ran `black tests/mussel/cli/test_tessellate.py`, all `OmegaConf.create({...})` calls are now properly formatted.

File: mussel/cli/tessellate_extract_features_common.py [hunk 2]
```diff
@@ -72,11 +73,28 @@ def _tessellate_and_filter(
...
+            seg_cfg.get("remove_penmarks"),
+        )
+
     if values := segment_tissue(
         slide_path=slide_path,
         slide_id=slide_id,
         output_h5_path=tessellate_h5_path,
-        **OmegaConf.to_container(cfg.seg_config),
+        artifact_remover_fn=artifact_remover_fn,
+        **seg_cfg,
```
[the remark this answers is not in the document]
Comment (raylim) [reply mussel/cli/tessellate_extract_features_common.py:96]: Fixed in 84bfb20. Wired `artifact_remover_fn` into all three CLI entrypoints:

- `tessellate.py::main()` — now builds the remover from `seg_cfg` and passes it to `segment_tissue()`
- `filter_tessellate.py::_main()` — same, also strips `artifact_exclude_classes` before spreading `seg_cfg`
- `tessellate_extract_features{_common}.py` — consolidated: `_build_artifact_remover` moved to `tessellate.py` (co-located with `SegConfig`), imported by both modules instead of duplicating the logic.

**Review by raylim — COMMENTED** [review]
The four bug fixes are correct and the new `exclude_classes` API is well-designed. A few minor issues below.

File: mussel/utils/segment.py [hunk 3]
```diff
@@ -757,7 +757,81 @@ def segment_tissue(
...
+                post_pixels = int(np.count_nonzero(result_mask))
+                removal_fraction = (
+                    1.0 - post_pixels / pre_pixels if pre_pixels > 0 else 1.0
+                )
+                if removal_fraction >= 1.0 - _MIN_TISSUE_SURVIVAL:
+                    logger.warning(
+                        "Artifact removal eliminated %.0f%% of tissue for slide %s "
+                        "(threshold: revert if >%.0f%% removed). "
+                        "Falling back to pre-removal mask. "
+                        "Consider using remove_penmarks_only=True or disabling "
```
Comment (raylim) [inline mussel/utils/segment.py]: Two minor issues in this warning message:

1. **Off-by-one in threshold description**: The condition is `removal_fraction >= 1.0 - _MIN_TISSUE_SURVIVAL` (i.e. `>= 0.90`), so the revert fires at *exactly* 90% removal too. The message should say `>=%.0f%%` not `>%.0f%%`.

2. **Deprecated parameter referenced**: `remove_penmarks_only=True` is the old (deprecated) parameter. The hint should use the current API:

```python
"(threshold: revert if >=%.0f%% removed). "
"Falling back to pre-removal mask. "
"Consider using artifact_exclude_classes=[4, 7] (pen marks only) "
"or disabling remove_artifacts for this slide type.",
```

**Review by raylim — COMMENTED** [review]
Logic and fixes all look correct. One minor comment on the warning message in `segment.py` — off-by-one in threshold description and a deprecated param reference. Everything else LGTM: the `>=` MPP escalation, tissue-survival fallback, DRY `_build_artifact_remover`, and wiring across all three CLIs are all solid. 22/22 tests pass.",

File: mussel/utils/segment.py (hunk 3 above)
Comment (raylim) [reply mussel/utils/segment.py]: Fixed in 4eaf7ee — changed `>` to `>=` in the threshold message and replaced `remove_penmarks_only=True` with `artifact_exclude_classes=[4, 7]`.
